### PR TITLE
Itertools are no longer reexported from util, optimized triedb iter

### PR DESCRIPTION
**Pull request by debris**
Triedb benches

old:

```
test trie_iter                      ... bench:   6,571,603 ns/iter (+/- 732,647)
```

new:

```
test trie_iter                      ... bench:   6,181,846 ns/iter (+/- 461,696)
```

### CODE REVIEW

File: util/src/trie/triedb.rs [hunk 1]
```diff
@@ -293,7 +292,17 @@ impl<'a> TrieDBIterator<'a> {
 	/// The present key.
 	fn key(&self) -> Bytes {
 		// collapse the key_nibbles down to bytes.
-		self.key_nibbles.iter().step(2).zip(self.key_nibbles.iter().skip(1).step(2)).map(|(h, l)| h * 16 + l).collect()
+		unsafe {
```
Comment (tomusdrw) [inline util/src/trie/triedb.rs]: Does it really require such big block of `unsafe` code to be fast?
what about:
```rust
let nibbles = self.key_nibbles;
let len = nibbles.len() - 1;
let mut i = 0;
while i < len {
  result.push(nibbles[i] * 16 + nibbles[i + 1]);
  i += 2;
}
```

Comment (rphmeier) [inline util/src/trie/triedb.rs]: maybe the checked indexing is too slow? i would guess that using unsafe only for `get_unchecked` would be good enough.

**Comment by rphmeier** [comment]
LGTM other than @tomusdrw grumble

File: util/src/trie/triedb.rs [hunk 2]
```diff
@@ -293,7 +292,15 @@ impl<'a> TrieDBIterator<'a> {
 	/// The present key.
 	fn key(&self) -> Bytes {
 		// collapse the key_nibbles down to bytes.
-		self.key_nibbles.iter().step(2).zip(self.key_nibbles.iter().skip(1).step(2)).map(|(h, l)| h * 16 + l).collect()
+		let nibbles = &self.key_nibbles;
+		let mut i = 0;
+		let mut result = Bytes::with_capacity(nibbles.len() / 2);
+		let len = nibbles.len() - 1;
```
Comment (arkpar) [inline util/src/trie/triedb.rs]: Is there a guarantee that `nibbles` is never empty and this can't underflow?

Comment (debris) [inline util/src/trie/triedb.rs]: no, just pushed a fix for that

**Pull request merged**